src/main/java/com/mraof/minestuck/world/gen/feature/structure/FrogTemplePiece.java

```diff
@@ -0,0 +1,348 @@
+	{
+		//lower room columns
+		for(int i = 0; i < 3; i++)
+		{
+			for(int j = 0; j < 3; j++)
+			{
+				fillWithBlocks(world, boundingBox, 10 + 20 + i * 10, 5, 48 + 20 + j * 10, 11 + 20 + i * 10, 23, 49 + 20 + j * 10, columnBlock.with(MSDirectionalBlock.FACING, Direction.UP), columnBlock.with(MSDirectionalBlock.FACING, Direction.UP), false);
+				fillWithBlocks(world, boundingBox, 8 + 20 + i * 10, 16, 46 + 20 + j * 10, 13 + 20 + i * 10, 16, 51 + 20 + j * 10, MSBlocks.CHISELED_GREEN_STONE_BRICKS.getDefaultState(), MSBlocks.CHISELED_GREEN_STONE_BRICKS.getDefaultState(), false); //platform around columns
+			}
+		}
+		
+		//lower room stepping stones
+		fillWithBlocks(world, boundingBox, 24 + 20, 16, 68 + 20, 24 + 20, 16, 69 + 20, MSBlocks.CHISELED_GREEN_STONE_BRICKS.getDefaultState(), MSBlocks.CHISELED_GREEN_STONE_BRICKS.getDefaultState(), false); //first step
+		fillWithBlocks(world, boundingBox, 30 + 20, 16, 65 + 20, 31 + 20, 16, 65 + 20, MSBlocks.CHISELED_GREEN_STONE_BRICKS.getDefaultState(), MSBlocks.CHISELED_GREEN_STONE_BRICKS.getDefaultState(), false); //second step
+		fillWithBlocks(world, boundingBox, 27 + 20, 16, 58 + 20, 27 + 20, 16, 59 + 20, MSBlocks.CHISELED_GREEN_STONE_BRICKS.getDefaultState(), MSBlocks.CHISELED_GREEN_STONE_BRICKS.getDefaultState(), false); //third step
+		fillWithBlocks(world, boundingBox, 17 + 20, 16, 58 + 20, 17 + 20, 16, 59 + 20, MSBlocks.CHISELED_GREEN_STONE_BRICKS.getDefaultState(), MSBlocks.CHISELED_GREEN_STONE_BRICKS.getDefaultState(), false); //fourth step
+		fillWithBlocks(world, boundingBox, 10 + 20, 16, 55 + 20, 11 + 20, 16, 55 + 20, MSBlocks.CHISELED_GREEN_STONE_BRICKS.getDefaultState(), MSBlocks.CHISELED_GREEN_STONE_BRICKS.getDefaultState(), false); //fifth step
+		fillWithBlocks(world, boundingBox, 14 + 20, 16, 48 + 20, 14 + 20, 16, 49 + 20, MSBlocks.CHISELED_GREEN_STONE_BRICKS.getDefaultState(), MSBlocks.CHISELED_GREEN_STONE_BRICKS.getDefaultState(), false); //sixth step
+		
+		//lower room ladder
+		fillWithBlocks(world, boundingBox, 20 + 20, 5, 72 + 20, 21 + 20, 16, 72 + 20, Blocks.LADDER.getDefaultState().with(LadderBlock.FACING, Direction.SOUTH), Blocks.LADDER.getDefaultState(), false);
+	}
+	
+	private void carveRooms(IWorld world, MutableBoundingBox boundingBox)
+	{
+		fillWithAir(world, boundingBox, 14 + 20, 49, 52 + 20, 27 + 20, 55, 65 + 20); //lotus room
+		fillWithAirCheckWater(world, boundingBox, 19 + 20, 49, 49 + 20, 22 + 20, 52, 51 + 20); //lotus room entry
+		fillWithAir(world, boundingBox, 7 + 20, 5, 45 + 20, 34 + 20, 23, 72 + 20); //lower room
+		fillWithAirCheckWater(world, boundingBox, 19 + 20, 17, 71 + 20, 22 + 20, 20, 79 + 20); //lower room entry //TODO this entrance will occasionally generate below water and cause this to generate awkwardly
+	}
+	
+	private void buildPillars(BlockState block, IWorld world, MutableBoundingBox boundingBox, Random random) //TODO pieces of pillars often fail to generate, which is why it is not in use currently
```

Comment:
Gonna repeat it here: I think it's a good idea to implement the pillars as separate structure pieces within the same structure